### PR TITLE
Qt: display game icon and title in the game list

### PR DESCRIPTION
**Pull request by wwylele**
list

@JamePeng told me this makes it much easier to find games.

### CODE REVIEW

**Comment by mailwl** [comment]
Nice, but "File type" column looks ugly now. Maybe replace it with images too, or shift between name and size?

File: src/common/common_funcs.h [hunk 1]
```diff
@@ -87,3 +87,7 @@ inline u64 _rotr64(u64 x, unsigned int shift){
 // This function might change the error code.
 // Defined in Misc.cpp.
 const char* GetLastErrorMsg();
+
+constexpr u32 MakeMagic(char a, char b, char c, char d) {
```
Comment (lioncash) [inline src/common/common_funcs.h]: This should likely be put elsewhere preferably in a place with a namespace. Ideally common_funcs shouldn't even exist as a header, since it pulls in a bunch of unrelated stuff when its included.

File: src/citra_qt/game_list_p.h [hunk 2]
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
...
+    GameListItemPath(const QString& game_path, const std::vector<u8>& smdh_data): GameListItem()
     {
         setData(game_path, FullPathRole);
+
+        if (smdh_data.size() < sizeof(SMDH)) return;
+        const SMDH* smdh = reinterpret_cast<const SMDH*>(smdh_data.data());
+        if (MakeMagic('S', 'M', 'D', 'H') != smdh->magic) return;
+
+        // Get icon from smdh
+        QImage icon(48, 48, QImage::Format::Format_RGB888);
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: Image retrieval should be its own function

File: src/citra_qt/game_list_p.h [hunk 3]
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
...
+                u32 coarse_y = y & ~7;
+                auto v = Color::DecodeRGB565(
+                    smdh->large_icon.data() + VideoCore::GetMortonOffset(x, y, 2) + coarse_y * 48 * 2);
+                icon.setPixel(x, y, qRgb(v.r(), v.g(), v.b()));
+            }
+        }
+        setData(QPixmap::fromImage(icon), Qt::DecorationRole);
+
+        // Get title form smdh
+        setData(QString::fromUtf16(smdh->titles[1].short_title.data()), TitleRole);
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: Title retrieval should be its own function.

File: src/citra_qt/game_list_p.h [hunk 4]
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
...
     static const int FullPathRole = Qt::UserRole + 1;
+    static const int TitleRole = Qt::UserRole + 2;
 
     GameListItemPath(): GameListItem() {}
-    GameListItemPath(const QString& game_path): GameListItem()
+    GameListItemPath(const QString& game_path, const std::vector<u8>& smdh_data): GameListItem()
     {
         setData(game_path, FullPathRole);
+
+        if (smdh_data.size() < sizeof(SMDH)) return;
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: Sanitizing should also be its own function (`IsValidSMDH or something')

File: src/citra_qt/game_list.cpp [hunk 5]
```diff
@@ -143,9 +145,17 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, bool d
                 LOG_WARNING(Frontend, "Filetype and extension of file %s do not match.", physical_name.c_str());
             }
 
+            std::vector<u8> smdh;
+            std::unique_ptr<Loader::AppLoader> loader;
+            if (filetype == Loader::FileType::CCI || filetype==Loader::FileType::CXI) {
```
Comment (lioncash) [inline src/citra_qt/game_list.cpp]: Acquiring the loader should be its own function as well.

File: src/citra_qt/game_list.cpp [hunk 6]
```diff
@@ -11,6 +11,8 @@
 #include "ui_settings.h"
 
 #include "core/loader/loader.h"
+#include "core/loader/ncch.h"
+#include "core/loader/3dsx.h"
```
Comment (lioncash) [inline src/citra_qt/game_list.cpp]: 3dsx.h comes before loader.h

File: src/citra_qt/game_list_p.h [hunk 7]
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
...
+        const SMDH* smdh = reinterpret_cast<const SMDH*>(smdh_data.data());
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: This reinterpret cast is undefined behavior.

File: src/core/loader/3dsx.cpp [hunk 8]
```diff
@@ -303,4 +303,29 @@ ResultStatus AppLoader_THREEDSX::ReadRomFS(std::shared_ptr<FileUtil::IOFile>& ro
     return ResultStatus::ErrorNotUsed;
 }
 
+ResultStatus AppLoader_THREEDSX::ReadIcon(std::vector<u8>& buffer) {
```
Comment (lioncash) [inline src/core/loader/3dsx.cpp]: You can use boost::optional here to avoid using an out parameter.

File: src/core/loader/ncch.cpp [hunk 9]
```diff
@@ -173,6 +173,11 @@ ResultStatus AppLoader_NCCH::LoadSectionExeFS(const char* name, std::vector<u8>&
     if (!file.IsOpen())
         return ResultStatus::Error;
 
+    if (!is_exefs_loaded) {
```
Comment (lioncash) [inline src/core/loader/ncch.cpp]: This should likely be its own function of some sort, it's copied twice.

File: src/core/loader/3dsx.cpp (hunk 8 above)
Comment (wwylele) [inline src/core/loader/3dsx.cpp]: It is a virtual function in `AppLoader` (though never used before). Should I still change it?

**Comment by wwylele** [comment]
default
Update: swap name and type column, and add a blank icon to those without SMDH data

**Comment by LittleWhite-tb** [comment]
Really nice feature :+1:

File: src/citra_qt/game_list_p.h [hunk 10]
```diff
@@ -5,13 +5,128 @@
...
+        int size = large ? 48 : 24;
+        QPixmap icon(size,size);
+        //Fill with white
+        icon.fill();
+        return icon;
+    }
+
+    ///an enum indicating the language for each title entry
+    enum class TitleLanguage {
+        JAPANESE = 0,
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: These should be the regularly cased. e.g. `Japanese`, `English`, etc.

File: src/citra_qt/game_list_p.h [hunk 11]
```diff
@@ -5,13 +5,128 @@
...
+     */
+    static QPixmap GetDefaultIcon(bool large) {
+        int size = large ? 48 : 24;
+        QPixmap icon(size,size);
+        //Fill with white
+        icon.fill();
+        return icon;
+    }
+
+    ///an enum indicating the language for each title entry
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: `/// Indicates the language used for each title entry`

The fact it's an enum is self-evident.

File: src/citra_qt/game_list_p.h [hunk 12]
```diff
@@ -5,13 +5,128 @@
...
+        RUSSIAN = 10,
+        TRADITIONAL_CHINESE = 11
+    };
+
+    /**
+     * Get game short title fromn SMDH
+     * @param language title language
+     * @return QString short title
+     */
+    QString GetShortTitle(TitleLanguage language) {
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: `Gets the short game title from the SMDH`

Also this function can be `const`

File: src/citra_qt/game_list_p.h [hunk 13]
```diff
@@ -5,13 +5,128 @@
...
+    }
+
+    /**
+     * Get default icon (for games without valid SMDH)
+     * @param large set true to get large icon (48x48), false to get small icon (24x24)
+     * @return QPixmap default icon
+     */
+    static QPixmap GetDefaultIcon(bool large) {
+        int size = large ? 48 : 24;
+        QPixmap icon(size,size);
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: Space after the comma

File: src/citra_qt/game_list_p.h [hunk 14]
```diff
@@ -5,13 +5,128 @@
...
+                auto v = Color::DecodeRGB565(
+                    icon_data + VideoCore::GetMortonOffset(x, y, 2) + coarse_y * size * 2);
+                icon.setPixel(x, y, qRgb(v.r(), v.g(), v.b()));
+            }
+        }
+        return QPixmap::fromImage(icon);
+    }
+
+    /**
+     * Get default icon (for games without valid SMDH)
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: `Gets the default icon (for games without a valid SMDH)`

File: src/citra_qt/game_list_p.h [hunk 15]
```diff
@@ -5,13 +5,128 @@
...
+     * @param large set true to get large icon (48x48), false to get small icon (24x24)
+     * @return QPixmap default icon
+     */
+    static QPixmap GetDefaultIcon(bool large) {
+        int size = large ? 48 : 24;
+        QPixmap icon(size,size);
+        //Fill with white
+        icon.fill();
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: Unless our background is explicitly set to white, this'll show a white block on OSes (or OS themes) that have a different background scheme

File: src/citra_qt/game_list_p.h [hunk 16]
```diff
@@ -5,13 +5,128 @@
...
+    std::array<u8, 0x480> small_icon;
+    std::array<u8, 0x1200> large_icon;
+
+    /**
+     * Test if data is a valid SMDH by its length and magic.
+     * @param smdh_data data buffer to test
+     * @return bool test result
+     */
+    static bool IsValidSMDH(const std::vector<u8>& smdh_data) {
+        if (smdh_data.size() < sizeof(SMDH)) return false;
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]:
```cpp
if (smdh_data.size() < sizeof(SMDH))
    return false;
```

File: src/citra_qt/game_list_p.h [hunk 17]
```diff
@@ -5,13 +5,128 @@
...
 #include <QString>
+#include <QImage>
+
+#include "core/loader/loader.h"
 
 #include "citra_qt/util/util.h"
 #include "common/string_util.h"
+#include "common/common_funcs.h"
+#include "common/common_types.h"
+#include "common/color.h"
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: color comes before common

File: src/citra_qt/game_list_p.h [hunk 18]
```diff
@@ -5,13 +5,128 @@
...
 
 #include <atomic>
+#include <array>
 
 #include <QRunnable>
 #include <QStandardItem>
 #include <QString>
+#include <QImage>
+
+#include "core/loader/loader.h"
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: core comes after common includes

File: src/citra_qt/game_list_p.h [hunk 19]
```diff
@@ -5,13 +5,128 @@
 #pragma once
 
 #include <atomic>
+#include <array>
 
 #include <QRunnable>
 #include <QStandardItem>
 #include <QString>
+#include <QImage>
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: This should be above QRunnable.

File: src/citra_qt/game_list_p.h [hunk 20]
```diff
@@ -27,29 +142,46 @@ class GameListItem : public QStandardItem {
...
+        if (!SMDH::IsValidSMDH(smdh_data)) {
+            // SMDH is not valid, set a default icon
+            setData(SMDH::GetDefaultIcon(true), Qt::DecorationRole);
+            return;
+        }
+
+        SMDH smdh;
+        memcpy(&smdh, smdh_data.data(), sizeof(SMDH));
+
+        // Get icon from smdh
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: SMDH

File: src/citra_qt/game_list_p.h [hunk 21]
```diff
@@ -27,29 +142,46 @@ class GameListItem : public QStandardItem {
...
+        setData(smdh.GetIcon(true), Qt::DecorationRole);
+
+        // Get title form smdh
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: SMDH

File: src/core/loader/loader.cpp [hunk 22]
```diff
@@ -90,6 +90,28 @@ const char* GetFileTypeString(FileType type) {
...
+
+    // 3DSX file format...
+    case FileType::THREEDSX:
+        return std::make_unique<AppLoader_THREEDSX>(std::move(file), filename, filepath);
+
+    // Standard ELF file format...
+    case FileType::ELF:
+        return std::make_unique<AppLoader_ELF>(std::move(file), filename);
+
+    // NCCH/NCSD container formats...
```
Comment (lioncash) [inline src/core/loader/loader.cpp]: the `...` are unnecessary

File: src/core/loader/loader.h [hunk 23]
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
 extern const std::initializer_list<Kernel::AddressMapping> default_address_mappings;
 
 /**
+ * Get a loader for a file with a specific type
+ * @param file The file to load
+ * @param type The type of the file
+ * @filename the file name (without path)
```
Comment (lioncash) [inline src/core/loader/loader.h]: `@param`

File: src/core/loader/loader.h [hunk 24]
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
...
+ * @filepath the file full path (with name)
```
Comment (lioncash) [inline src/core/loader/loader.h]: @param

File: src/core/loader/loader.h [hunk 25]
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
 extern const std::initializer_list<Kernel::AddressMapping> default_address_mappings;
 
 /**
+ * Get a loader for a file with a specific type
```
Comment (lioncash) [inline src/core/loader/loader.h]: There should be a blank newline after this to separate the description from the parameter documentation

File: src/core/loader/ncch.cpp [hunk 26]
```diff
@@ -173,6 +173,9 @@ ResultStatus AppLoader_NCCH::LoadSectionExeFS(const char* name, std::vector<u8>&
     if (!file.IsOpen())
         return ResultStatus::Error;
 
+    ResultStatus r = LoadExeFS();
+    if (r != ResultStatus::Success) return r;
```
Comment (lioncash) [inline src/core/loader/ncch.cpp]:
```cpp
if (r != ResultStatus::Success)
    return r;
```

File: src/core/loader/ncch.cpp [hunk 27]
```diff
@@ -282,6 +285,17 @@ ResultStatus AppLoader_NCCH::Load() {
...
+    is_exefs_loaded = true;
+    return ResultStatus::Success;
+}
+
+ResultStatus AppLoader_NCCH::Load() {
+    if (is_loaded)
+        return ResultStatus::ErrorAlreadyLoaded;
+
+    ResultStatus r = LoadExeFS();
+    if (r != ResultStatus::Success) return r;
```
Comment (lioncash) [inline src/core/loader/ncch.cpp]:
```cpp
if (r != ResultStatus::Success)
    return r;
```

File: src/core/loader/loader.h (hunk 25 above)
Comment (wwylele) [inline src/core/loader/loader.h]: @lioncash all other documentation around don't have blank line separating the description from the parameter... should I still add it? and to all other documentation?

**Comment by wwylele** [comment]
@lioncash Thank you very much for reviewing! By the way, I think the [Contributing guide](https://github.com/citra-emu/citra/blob/master/CONTRIBUTING.md) needs update. Something there is unclear ( e.g. the naming rule for enum).

File: src/core/loader/loader.h (hunk 25 above)
Comment (lioncash) [inline src/core/loader/loader.h]: Oh in that case don't worry about it, then.

File: src/citra_qt/game_list_p.h [hunk 28]
```diff
@@ -5,13 +5,128 @@
 #pragma once
 
 #include <atomic>
+#include <array>
```
Comment (lioncash) [inline src/citra_qt/game_list_p.h]: Sorry, I missed this one, this should be above `<atomic>`

**Comment by mailwl** [comment]
@wwylele, may i ask you as Qt-guru create toolbar with all this thigger buttons, which now in "Configure.." menu button? I hate this redundant clicks to get needed option with new Citra Configuration window.

Sorry for offtop

**Comment by wwylele** [comment]
@mailwl technically it's not hard, but I don't know if everyone like that... if you really want it, I can make it for you on my branch someday.

**Comment by mailwl** [comment]
hmm... maybe add to menu add option to show/hide toolbar? I think, this will be useful even in future release state of Citra.
eh, it's not the place for discussion offtopics

File: src/citra_qt/game_list.cpp [hunk 29]
```diff
@@ -34,8 +34,8 @@ GameList::GameList(QWidget* parent)
     tree_view->setUniformRowHeights(true);
 
     item_model->insertColumns(0, COLUMN_COUNT);
-    item_model->setHeaderData(COLUMN_FILE_TYPE, Qt::Horizontal, "File type");
     item_model->setHeaderData(COLUMN_NAME, Qt::Horizontal, "Name");
```
Comment (JayFoxRox) [inline src/citra_qt/game_list.cpp]: Can you tell this column to use the available space or something? The default settings for me were that it only had enough room for the icon, I couldn't even see the games name.

Comment (wwylele) [inline src/citra_qt/game_list.cpp]: @JayFoxRox added, but not sure if it's in a correct way.

File: src/citra_qt/game_list.cpp [hunk 30]
```diff
@@ -143,9 +147,12 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, bool d
                 LOG_WARNING(Frontend, "Filetype and extension of file %s do not match.", physical_name.c_str());
             }
 
+            std::vector<u8> smdh;
+            std::unique_ptr<Loader::AppLoader> loader = Loader::GetLoader(FileUtil::IOFile(physical_name, "rb"), filetype, filename_filename, physical_name);
+            if (loader) loader->ReadIcon(smdh);
```
Comment (bunnei) [inline src/citra_qt/game_list.cpp]: new line

File: src/citra_qt/game_list.cpp [hunk 31]
```diff
@@ -143,9 +147,12 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, bool d
...
             emit EntryReady({
```
Comment (bunnei) [inline src/citra_qt/game_list.cpp]: Could use a space before this and the statements above

File: src/core/loader/ncch.h [hunk 32]
```diff
@@ -232,6 +232,14 @@ class AppLoader_NCCH final : public AppLoader {
...
     ResultStatus LoadExec();
 
+    /**
+     * Ensure ExeFS is loaded and ready for reading sections
+     * @return ResultStatus result of function
+     */
+    ResultStatus LoadExeFS();
+
+    bool            is_exefs_loaded = false;
+
```
Comment (bunnei) [inline src/core/loader/ncch.h]: Unnecessary extra space

File: src/core/loader/ncch.cpp [hunk 33]
```diff
@@ -282,6 +286,18 @@ ResultStatus AppLoader_NCCH::Load() {
...
 
+    is_exefs_loaded = true;
+    return ResultStatus::Success;
+}
+
+ResultStatus AppLoader_NCCH::Load() {
+    if (is_loaded)
+        return ResultStatus::ErrorAlreadyLoaded;
+
+    ResultStatus r = LoadExeFS();
```
Comment (bunnei) [inline src/core/loader/ncch.cpp]: Use `result` instead of `r`. Single letter variable names are easily abused and are against our style guide, unless for a loop index e.g. i

File: src/core/loader/ncch.cpp [hunk 34]
```diff
@@ -173,6 +173,10 @@ ResultStatus AppLoader_NCCH::LoadSectionExeFS(const char* name, std::vector<u8>&
     if (!file.IsOpen())
         return ResultStatus::Error;
 
+    ResultStatus r = LoadExeFS();
```
Comment (bunnei) [inline src/core/loader/ncch.cpp]: Ditto on single letter variable name

File: src/core/loader/loader.h [hunk 35]
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
...
+ * @param file The file to load
+ * @param type The type of the file
+ * @param filename the file name (without path)
+ * @param filepath the file full path (with name)
+ * @return std::unique_ptr<AppLoader> a pointer to a loader object;  null pointer for unsupported type
+ */
+std::unique_ptr<AppLoader> GetLoader(FileUtil::IOFile&& file, FileType type,
+    const std::string& filename, const std::string& filepath);
```
Comment (bunnei) [inline src/core/loader/loader.h]: Unnecessary line break here

File: src/core/loader/loader.h [hunk 36]
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
 extern const std::initializer_list<Kernel::AddressMapping> default_address_mappings;
 
 /**
+ * Get a loader for a file with a specific type
+ * @param file The file to load
+ * @param type The type of the file
+ * @param filename the file name (without path)
+ * @param filepath the file full path (with name)
+ * @return std::unique_ptr<AppLoader> a pointer to a loader object;  null pointer for unsupported type
```
Comment (bunnei) [inline src/core/loader/loader.h]: *nullptr

File: src/core/loader/3dsx.cpp [hunk 37]
```diff
@@ -303,4 +303,29 @@ ResultStatus AppLoader_THREEDSX::ReadRomFS(std::shared_ptr<FileUtil::IOFile>& ro
...
+        return ResultStatus::Error;
+
+    if (hdr.header_size != sizeof(THREEDSX_Header))
+        return ResultStatus::Error;
+
+    // Check if the 3DSX has a SMDH...
+    if (hdr.smdh_offset != 0) {
+        file.Seek(hdr.smdh_offset, SEEK_SET);
+        buffer.resize(hdr.smdh_size);
+        if (file.ReadBytes(&buffer[0], hdr.smdh_size) != hdr.smdh_size)
```
Comment (bunnei) [inline src/core/loader/3dsx.cpp]: Space before this line

File: src/core/loader/3dsx.cpp [hunk 38]
```diff
@@ -303,4 +303,29 @@ ResultStatus AppLoader_THREEDSX::ReadRomFS(std::shared_ptr<FileUtil::IOFile>& ro
...
+            return ResultStatus::Error;
```
Comment (bunnei) [inline src/core/loader/3dsx.cpp]: Space after this line. Don't make your code so crammed!

File: src/citra_qt/game_list_p.h [hunk 39]
```diff
@@ -27,29 +142,46 @@ class GameListItem : public QStandardItem {
...
-        if (role == FullPathRole) {
+    QVariant data(int role) const override {
+        if (role == Qt::DisplayRole) {
             std::string filename;
-            Common::SplitPath(value.toString().toStdString(), nullptr, &filename, nullptr);
-            GameListItem::setData(QString::fromStdString(filename), Qt::DisplayRole);
-            GameListItem::setData(value, FullPathRole);
+            Common::SplitPath(data(FullPathRole).toString().toStdString(), nullptr, &filename, nullptr);
+            QString title = data(TitleRole).toString();
+            if (title.isEmpty())
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: Use a ternary.. e.g. `return title.isEmpty() ? foo : bar;`

File: src/citra_qt/game_list_p.h [hunk 40]
```diff
@@ -4,14 +4,129 @@
...
+        if (smdh_data.size() < sizeof(SMDH))
+            return false;
+        u32 magic;
+        memcpy(&magic, smdh_data.data(), 4);
+        return Loader::MakeMagic('S', 'M', 'D', 'H') == magic;
+    }
+
+    /**
+     * Gets game icon from SMDH
+     * @param large set true to get large icon (48x48), false to get small icon (24x24)
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: Grammar nit:
"If true, returns large icon (48x48), otherwise returns small icon (24x24)"

File: src/citra_qt/game_list_p.h [hunk 41]
```diff
@@ -4,14 +4,129 @@
...
+    INSERT_PADDING_BYTES(2);
+    float_le banner_animation_frame;
+    u32_le cec_id;
+    INSERT_PADDING_BYTES(8);
+
+    std::array<u8, 0x480> small_icon;
+    std::array<u8, 0x1200> large_icon;
+
+    /**
+     * Tests if data is a valid SMDH by its length and magic.
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: "magic number"

File: src/citra_qt/game_list_p.h [hunk 42]
```diff
@@ -4,14 +4,129 @@
...
+     * @param smdh_data data buffer to test
+     * @return bool test result
+     */
+    static bool IsValidSMDH(const std::vector<u8>& smdh_data) {
+        if (smdh_data.size() < sizeof(SMDH))
+            return false;
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: Space after this line

File: src/citra_qt/game_list_p.h [hunk 43]
```diff
@@ -4,14 +4,129 @@
...
+        u32 magic;
+        memcpy(&magic, smdh_data.data(), 4);
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: Space after this line

File: src/citra_qt/game_list_p.h [hunk 44]
```diff
@@ -4,14 +4,129 @@
...
+#include "common/color.h"
+#include "common/common_funcs.h"
+#include "common/common_types.h"
+
+#include "core/loader/loader.h"
+
+#include "video_core/utils.h"
+
+/// SMDH data structure that contains title, titles etc. See https://www.3dbrew.org/wiki/SMDH
+struct SMDH {
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: This code to decode SMDH icons does not belong in the Qt frontend, it should be in core since it's part of 3DS emulation. Probably belongs in core/loader/ncch.*, since this icon format is specific to CCI/CXI.

**Comment by bunnei** [comment]
Code generally looks fine, aside from the nits and the restructure to move icon decoding to core/loader. Some other general comments:
* I don't see the need to print filenames in the list view, I would just print the game name. We probably will (at some point) want to have an internal DB of nice names, but for now I think just doing this is much nicer than file names.
* I already noticed I have some duplicate filenames in my library for games that I have multiple regions of. Not critical for this PR, but if you're interested in continuing to work on the UI, it would be awesome to also have a region column in the game list view.

Anyway, this looks awesome and I hope to get it merged soon!

**Comment by wwylele** [comment]
Addressed @bunnei 's comment. And:
- I moved SMDH structure to Loader.h because 3dsx also contains data in that format. Also for reducing header including in game_list_p.h
- After this PR merged I will work on adding region info. (we are really getting identical to dolphin's game list :sweat_smile: )
- I want to keep file names, at least for now, for the following reason:
 -  It is useful as an extra description for each game file. For instance, I give some games file name with region and game ID, so that I can distinguish same game from different region, and get ID easily when I need it.
 - The name column of game list is currently functioning not only for displaying, but also for storing file path for booting. It's a little bit weird to store file name but not to display them. So it needs some reconstruction if it doesn't display file name. I want to leave the for next PR (maybe along with region display)

**Comment by wwylele** [comment]
Found game_list_p.h was missing in CMakeList, so I added it. Sorry for strange commit order - my computer's system clock just screwed up.

File: src/citra_qt/game_list.cpp [hunk 45]
```diff
@@ -109,7 +109,11 @@ void GameList::SaveInterfaceLayout()
 void GameList::LoadInterfaceLayout()
 {
     auto header = tree_view->header();
-    header->restoreState(UISettings::values.gamelist_header_state);
+    if (!header->restoreState(UISettings::values.gamelist_header_state)) {
+        // We are using the name column to display icons and titles
+        // so make it as large as possible as default.
+        header->resizeSection(COLUMN_NAME, header->width());
```
Comment (MerryMage) [inline src/citra_qt/game_list.cpp]: You should likely use [`setSectionResizeMode`](http://doc.qt.io/qt-5/qheaderview.html#setSectionResizeMode) with `QHeaderView::Stretch`.

Comment (wwylele) [inline src/citra_qt/game_list.cpp]: @MerryMage If I set it to `Stretch`, user won't be able to resize the column.

File: src/core/loader/3dsx.cpp (hunk 8 above)
Comment (bunnei) [inline src/core/loader/3dsx.cpp]: I think it's fine, refactoring the interface (which is used elsewhere in AppLoader) is out of scope of this PR, IMO

File: src/citra_qt/game_list_p.h [hunk 46]
```diff
@@ -27,29 +99,43 @@ class GameListItem : public QStandardItem {
  * A specialization of GameListItem for path values.
  * This class ensures that for every full path value it holds, a correct string representation
  * of just the filename (with no extension) will be displayed to the user.
+ * If this class recieves a valid SMDH data, it will also display game icons and titles.
```
Comment (bunnei) [inline src/citra_qt/game_list_p.h]: *receives valid (no 'a')

**Comment by bunnei** [comment]
Code looks fine to me.

Last call for additional reviews on this one.

**Comment by Subv** [comment]
LGTM

**Comment by bunnei** [comment]
LGTM as well, ofc
